### PR TITLE
Fix S2259: "Null pointer dereference" raises FP when a variable is in…

### LINKED ISSUE #493: Fix S2259: "Null pointer dereference" raises FP when a variable is initialized in catch(Exception) block
```
object o = null;
try
{
    o = new object();
    // do something
}
catch
{
    o = new object();
}
return o.ToString(); // FP is detected
```

TIP: The CFG has connection between the beginning of the try block and its end for the cases when a potential exception is not caught. However, catch or catch(Exception) will catch all exceptions, meaning that the connection is not needed in this case.

### PR DESCRIPTION
**Pull request by valhristov**
…itialized in catch(Exception) block

Fix #493

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SyntaxHelper.cs [hunk 1]
```diff
@@ -202,5 +202,18 @@ public static int GetLineNumber(this SyntaxToken token)
         {
             return token.GetLocation().GetLineSpan().StartLinePosition.Line;
         }
+
+        public static bool IsCatchException(this CatchClauseSyntax catchClause)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SyntaxHelper.cs]: Split this into `IsEmptyCatch` and `IsCatchingRootException`/`IsCatchingAllExceptions`

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs [hunk 2]
```diff
@@ -532,6 +532,48 @@ void TestNameOf()
             var y = nameof(a, b);
             x.ToString();
         }
+
+        string TryCath1()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs]: Typo

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs [hunk 3]
```diff
@@ -532,6 +532,48 @@ void TestNameOf()
...
+                o = new object();
+            }
+            catch
+            {
+                o = new object();
+            }
+            return o.ToString();
+        }
+
+        string TryCath2()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs]: Typo

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs [hunk 4]
```diff
@@ -532,6 +532,48 @@ void TestNameOf()
...
+                o = new object();
+            }
+            catch (Exception)
+            {
+                o = new object();
+            }
+            return o.ToString();
+        }
+
+        string TryCath3()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereference.cs]: Typo

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereferenceCSharp6.cs [hunk 5]
```diff
@@ -13,5 +13,19 @@ void ConditionalThisFieldAccess()
             this._foo1 = o;
             this?._foo1.ToString(); // Noncompliant
         }
+
+        string TryCath3()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NullPointerDereferenceCSharp6.cs]: Typo

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SyntaxHelper.cs (hunk 1 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SyntaxHelper.cs]: I think I'm missing something - why split? Wouldn't renaming to IsCatchingAllExceptions (I like the name!) be sufficient?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 6]
```diff
@@ -479,39 +479,69 @@ private void BuildLabeledStatement(LabeledStatementSyntax labeledStatement)
...
-            foreach (var catchClause in tryStatement.Catches.Reverse())
-            {
-                currentBlock = CreateBlock(finallyBlock);
+            var catchBlocks = BuildCatchClauses(tryStatement.Catches.Reverse(), catchSuccessor);
 
-                BuildBlock(catchClause.Block);
+            // If there is a catch with no Exception filter or equivalent we don't want to
+            // join the tryStatement start/end blocks with the exit block because all
+            // exceptions will be caught before going to finally
+            var catchesAllExceptions = tryStatement.Catches.Any(SyntaxHelper.IsCatchException);
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: Rename to isCatchingAllExceptions

**Pull request merged**